File: src/main/java/com/baulsupp/oksocial/lyft/LyftAuthFlow.java
package com.baulsupp.oksocial.lyft;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.util.Map;
import okhttp3.Credentials;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class LyftAuthFlow {
  public static LyftServerCredentials login(OkHttpClient client) throws IOException {
    String clientId = System.console().readLine("Lyft Client Id: ");
    String clientSecret = new String(System.console().readPassword("Lyft Client Secret: "));

    RequestBody body = RequestBody.create(MediaType.parse("application/json"),
        "{\"grant_type\": \"client_credentials\", \"scope\": \"public\"}");
    String basic = Credentials.basic(clientId, clientSecret);
    Request request =
        new Request.Builder().url("https://api.lyft.com/oauth/token")
            .post(body)
            .header("Authorization", basic)
            .build();

    Response response = client.newCall(request).execute();

    try {
      if (response.code() != 200) {
        throw new IOException("Failed auth " + response.code() + response.message());
      } else {
        ObjectMapper mapper = new ObjectMapper();
        Map<String, String> map =
            mapper.readValue(response.body().string(), new TypeReference<Map<String, String>>() {
            });

        return new LyftServerCredentials(map.get("access_token"));
      }
    } finally {
      response.body().close();
    }
  }
}

File: src/main/kotlin/com/baulsupp/okurl/services/lyft/LyftAuthFlow.kt
package com.baulsupp.okurl.services.lyft

import com.baulsupp.oksocial.output.OutputHandler
import com.baulsupp.okurl.authenticator.SimpleWebServer
import com.baulsupp.okurl.authenticator.oauth2.Oauth2Token
import com.baulsupp.okurl.credentials.NoToken
import com.baulsupp.okurl.kotlin.queryMap
import com.baulsupp.okurl.kotlin.requestBuilder
import okhttp3.Credentials
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.net.URLEncoder

object LyftAuthFlow {
  suspend fun login(
    client: OkHttpClient,
    outputHandler: OutputHandler<Response>,
    clientId: String,
    clientSecret: String,
    scopes: Iterable<String>
  ): Oauth2Token {
    SimpleWebServer.forCode().use { s ->
      val scopesString = URLEncoder.encode(scopes.joinToString(" "), "UTF-8")

      val loginUrl =
        "https://api.lyft.com/oauth/authorize?client_id=$clientId&response_type=code&scope=$scopesString&state=x"

      outputHandler.openLink(loginUrl)

      val code = s.waitForCode()

      val body = "{\"grant_type\": \"authorization_code\", \"code\": \"$code\"}"
        .toRequestBody("application/json".toMediaType())
      val basic = Credentials.basic(clientId, clientSecret)
      val request = requestBuilder(
        "https://api.lyft.com/oauth/token",
        NoToken
      )
        .post(body)
        .header("Authorization", basic)
        .build()

      val responseMap = client.queryMap<String>(request)

      return Oauth2Token(
        responseMap["access_token"] as String,
        responseMap["refresh_token"] as String, clientId, clientSecret
      )
    }
  }
}
